CHANGES_UNRELEASED.md

```diff
@@ -4,6 +4,16 @@
 
 [Full Cihangelog](https://github.com/mozilla/application-services/compare/v0.30.0...master)
 
+## FxA Client
+
+### What's new
+
+- Better handling of invalid access tokens: when handling an authentication exception
+  thrown while using a token obtained through `FirefoxAccount.getAccessToken`,
+  applications should call `FirefoxAccount.onAccessTokenError` then retry one more time.
+  On a second failure they should begin a new OAuth flow.
+  `FirefoxAccount.getProfile` calls `onAccessTokenError` and retries once automatically.
```

Comment:
@rfk having implemented the suggested flow, I think we can do much better at the API level. We've talked about adding a function to check connectivity and return a boolean, let's do that. Otherwise, the current API leaks a lot of the internal details, and requires quite a lot of comments at the point of integration to explain what's going on. For example, there's no mention of "access token", or "tokens" in general at all in the a-c account manager, until we get to the "try to recover" bits, at which point there are detailed comment blocks describing refresh tokens and other fxalib internals. Things will be easier to understand if this was hidden behind a "ableToConnect: bool" method of sorts.

Reply:
We should probably combine these two methods. `onAccessTokenError` can return a boolean flag indicating if re-auth needs to happen, and clear its internal state. Well behaving clients should call both of these calls in quick succession anyway, so might as well combine them and lessen the cognitive burden.